src/sema/statements.rs

```diff
@@ -724,15 +727,33 @@ fn statement(
 
                         res.push(Statement::Delete(*loc, ty.as_ref().clone(), expr));
 
-                        Ok(true)
+                        return Ok(true);
                     } else {
+                        // For memory arrays and other types, we should only delete the element
+                        // by assigning the default value, not delete the entire array
+                        // Issue #1785 - match Solc behavior for delete array[index]
                         ns.diagnostics.push(Diagnostic::warning(
                             *loc,
                             "argument to 'delete' should be storage reference".to_string(),
                         ));
 
-                        Err(())
-                    };
+                        let expr_ty = expr.ty().clone();
+                        let element_ty = expr_ty.deref_any();
+                        
+                        if let Ok(default_expr) = get_default_value(*loc, element_ty, ns, diagnostics) {
+                            let assign = Expression::Assign {
+                                loc: *loc,
+                                ty: expr.ty(),
+                                left: Box::new(expr),
+                                right: Box::new(default_expr),
+                            };
+                            
+                            res.push(Statement::Expression(*loc, true, assign));
```

Comment:
The problem here is that the AST should be an accurate representation of the source code, not of the code it will generate - that happens in codegen.

So in sema, add:
```rust
res.push(Statement::Delete(...);
```
Then in codegen, for array elements, set a default value, here:

https://github.com/hyperledger-solang/solang/blob/main/src/codegen/statements/mod.rs#L217